File: test_bidi.cpp
// Standalone test for the bidi reordering function
// Compile with: g++ -std=c++17 -I/usr/include/x86_64-linux-gnu/qt6 -I/usr/include/qt6 -fPIC test_bidi.cpp -lQt6Core -o test_bidi
#include <QString>
#include <QVector>
#include <QChar>
#include <iostream>

// --- Copy of the bidi reordering functions from generator_pdf.cpp ---

static bool isRtlChar(QChar c)
{
    const QChar::Direction d = c.direction();
    return d == QChar::DirR || d == QChar::DirAL;
}

static bool hasRtlChars(const QString &str)
{
    for (int i = 0; i < str.length(); i++) {
        if (isRtlChar(str.at(i))) {
            return true;
        }
    }
    return false;
}

static bool isPrimaryRtl(const QString &str)
{
    for (int i = 0; i < str.length(); i++) {
        const QChar::Direction d = str.at(i).direction();
        if (d == QChar::DirR || d == QChar::DirAL) {
            return true;
        }
        if (d == QChar::DirL) {
            return false;
        }
    }
    return false;
}

struct BidiResult {
    QString text;
    QVector<int> logicalToVisual;
};

static BidiResult reorderVisualToLogical(const QString &visual)
{
    BidiResult result;
    const int len = visual.length();

    if (len <= 1 || !hasRtlChars(visual)) {
        result.text = visual;
        result.logicalToVisual.resize(len);
        for (int i = 0; i < len; i++) {
            result.logicalToVisual[i] = i;
        }
        return result;
    }

    const bool primaryRtl = isPrimaryRtl(visual);
    QString reordered;
    reordered.reserve(len);
    QVector<int> mapping;
    mapping.reserve(len);

    if (!primaryRtl) {
        int i = 0;
        while (i < len) {
            while (i < len && !isRtlChar(visual.at(i))) {
                reordered += visual.at(i);
                mapping.append(i);
                i++;
            }
            int j = i;
            while (j < len && isRtlChar(visual.at(j))) {
                j++;
            }
            if (j > i) {
                for (int k = j - 1; k >= i; k--) {
                    reordered += visual.at(k);
                    mapping.append(k);
                }
                i = j;
            }
        }
    } else {
        int i = len - 1;
        while (i >= 0) {
            while (i >= 0 && isRtlChar(visual.at(i))) {
                reordered += visual.at(i);
                mapping.append(i);
                i--;
            }
            int j = i;
            while (j >= 0 && !isRtlChar(visual.at(j))) {
                j--;
            }
            if (j < i) {
                for (int k = j + 1; k <= i; k++) {
                    reordered += visual.at(k);
                    mapping.append(k);
                }
                i = j;
            }
        }
    }

    result.text = reordered;
    result.logicalToVisual = mapping;
    return result;
}

// --- Test runner ---

struct TestCase {
    QString name;
    QString input;       // visual order
    QString expected;    // expected logical order
};

int main() {
    QVector<TestCase> tests = {
        // Test 1: Pure English - no change
        {"English LTR", "Hello World", "Hello World"},

        // Test 2: Pure Persian (Persian word "ketab" = book: ک ت ا ب)
        // Visual order: leftmost=ب, rightmost=ک
        // Expected logical: ک ت ا ب (right-to-left)
        {"Persian word",
         QString::fromUtf8("\xD8\xA8\xD8\xAA\xD8\xA7\xDA\xA9"),  // visual: ب ت ا ک
         QString::fromUtf8("\xDA\xA9\xD8\xA7\xD8\xAA\xD8\xA8")}, // logical: ک ا ت ب

        // Test 3: Mixed Persian + English ("Hello سلام")
        // Visual order: H e l l o   س ل ا م
        // Expected: Hello م ا ل س (LTR then RTL)
        {"Mixed Persian+English",
         QString::fromUtf8("Hello ") + QString::fromUtf8("\xD8\xB3\xD9\x84\xD8\xA7\xD9\x85"),
         QString::fromUtf8("Hello ") + QString::fromUtf8("\xD9\x85\xD8\xA7\xD9\x84\xD8\xB3")},

        // Test 4: Single RTL char - no change needed
        {"Single RTL char",
         QString::fromUtf8("\xD8\xB3"),  // س
         QString::fromUtf8("\xD8\xB3")},

        // Test 5: Hebrew word שלום (shalom)
        {"Hebrew word",
         QString::fromUtf8("\xD7\x9D\xD7\x95\xD7\x9C\xD7\xA9"),  // visual: ם ו ל ש
         QString::fromUtf8("\xD7\xA9\xD7\x9C\xD7\x95\xD7\x9D")}, // logical: ש ל ו ם

        // Test 6: Empty string
        {"Empty string", "", ""},
    };

    int passed = 0;
    int failed = 0;

    for (const auto &test : tests) {
        BidiResult result = reorderVisualToLogical(test.input);

        bool ok = (result.text == test.expected);
        if (ok) {
            passed++;
            std::cout << "PASS: " << test.name.toStdString() << std::endl;
        } else {
            failed++;
            std::cout << "FAIL: " << test.name.toStdString() << std::endl;
            std::cout << "  Input:    " << std::hex;
            for (int i = 0; i < test.input.length(); i++) {
                std::cout << "U+" << test.input.at(i).unicode() << " ";
            }
            std::cout << std::dec << std::endl;
            std::cout << "  Expected: ";
            for (int i = 0; i < test.expected.length(); i++) {
                std::cout << "U+" << std::hex << test.expected.at(i).unicode() << " ";
            }
            std::cout << std::dec << std::endl;
            std::cout << "  Got:      ";
            for (int i = 0; i < result.text.length(); i++) {
                std::cout << "U+" << std::hex << result.text.at(i).unicode() << " ";
            }
            std::cout << std::dec << std::endl;

            // Show mapping
            std::cout << "  Mapping:  ";
            for (int i = 0; i < result.logicalToVisual.size(); i++) {
                std::cout << result.logicalToVisual[i] << " ";
            }
            std::cout << std::endl;
        }
    }

    std::cout << "\n---" << std::endl;
    std::cout << "Results: " << passed << "/" << (passed + failed) << " passed" << std::endl;

    return failed > 0 ? 1 : 0;
}

File: test_bidi.py
#!/usr/bin/env python3
"""Test the bidi reordering algorithm (Python version of C++ code)"""

import unicodedata

def is_rtl_char(c):
    """Check if character has RTL direction"""
    bidi = unicodedata.bidirectional(c)
    return bidi in ('R', 'AL')  # Right-to-Left, Arabic Letter

def is_strong_ltr(c):
    bidi = unicodedata.bidirectional(c)
    return bidi == 'L'

def has_rtl_chars(s):
    return any(is_rtl_char(c) for c in s)

def is_primary_rtl(s):
    for c in s:
        bidi = unicodedata.bidirectional(c)
        if bidi in ('R', 'AL'):
            return True
        if bidi == 'L':
            return False
    return False

def reorder_visual_to_logical(visual):
    """Reorder from visual to logical order"""
    if len(visual) <= 1 or not has_rtl_chars(visual):
        return visual, list(range(len(visual)))
    
    primary_rtl = is_primary_rtl(visual)
    result = []
    mapping = []
    n = len(visual)
    
    if not primary_rtl:
        i = 0
        while i < n:
            while i < n and not is_rtl_char(visual[i]):
                result.append(visual[i])
                mapping.append(i)
                i += 1
            j = i
            while j < n and is_rtl_char(visual[j]):
                j += 1
            if j > i:
                for k in range(j - 1, i - 1, -1):
                    result.append(visual[k])
                    mapping.append(k)
                i = j
    else:
        i = n - 1
        while i >= 0:
            while i >= 0 and is_rtl_char(visual[i]):
                result.append(visual[i])
                mapping.append(i)
                i -= 1
            j = i
            while j >= 0 and not is_rtl_char(visual[j]):
                j -= 1
            if j < i:
                for k in range(j + 1, i + 1):
                    result.append(visual[k])
                    mapping.append(k)
                i = j
    
    return ''.join(result), mapping


def run_tests():
    tests = [
        # (name, visual_input, expected_logical, expected_mapping)
        ("English LTR", "Hello World", "Hello World", [0,1,2,3,4,5,6,7,8,9,10]),
        
        # Persian word "ketab" (کتاب): in PDF visual order = ب ا ت ک
        # Logical order = ک ت ا ب
        ("Persian word", "باتک", "کتاب", None),
        
        # Persian "salam" (سلام): in PDF visual = م ا ل س
        # Logical = س ل ا م
        ("Persian salam", "مالس", "سلام", None),
        
        # Mixed: "Hello سلام"
        # Visual: H e l l o _ س ل ا م
        # Logical: Hello _ م ا ل س
        ("Mixed EN+FA", "Hello سلام", "Hello م\u200fا\u200fل\u200fس", None),
        
        # Hebrew "shalom" (שלום): visual = ם ו ל ש
        # Logical = ש ל ו ם
        ("Hebrew shalom", "םולש", "שלום", None),
        
        # Single char
        ("Single RTL", "س", "س", [0]),
        
        # Empty
        ("Empty", "", "", []),
    ]
    
    passed = 0
    failed = 0
    
    for name, visual, expected, expected_map in tests:
        result, mapping = reorder_visual_to_logical(visual)
        
        # For mixed text, strip control chars for comparison
        clean_result = result.replace('\u200f', '').replace('\u200e', '')
        clean_expected = expected.replace('\u200f', '').replace('\u200e', '')
        
        ok = (clean_result == clean_expected)
        if ok:
            passed += 1
            print(f"PASS: {name}")
            print(f"  Visual:  {visual!r} -> Logical: {result!r}")
            print(f"  Mapping: {mapping}")
        else:
            failed += 1
            print(f"FAIL: {name}")
            print(f"  Visual:    {visual!r}")
            print(f"  Visual hex: {' '.join(f'U+{ord(c):04X}' for c in visual)}")
            print(f"  Expected:  {expected!r}")
            print(f"  Got:       {result!r}")
            print(f"  Got hex:   {' '.join(f'U+{ord(c):04X}' for c in result)}")
            print(f"  Mapping:   {mapping}")
        print()
    
    print(f"{'='*50}")
    print(f"Results: {passed}/{passed+failed} passed")
    return failed == 0


if __name__ == "__main__":
    success = run_tests()
    exit(0 if success else 1)
